Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,21 +14,279 @@
         config: KafkaConfig = KafkaConfig(),
         topicConfig: KafkaTopicConfig = KafkaTopicConfig(),
         logger: Logger
-    ) throws {
-        self.client = try KafkaClient(type: .producer, config: config, logger: logger)
+    ) async throws {
+        self.config = config
         self.topicConfig = topicConfig
+        self.logger = logger
+        self.topicHandles = [:]
+        self.state = .started
+
+        // (NIOAsyncSequenceProducer.makeSequence Documentation Excerpt)
+        // This method returns a struct containing a NIOAsyncSequenceProducer.Source and a NIOAsyncSequenceProducer.
+        // The source MUST be held by the caller and used to signal new elements or finish.
+        // The sequence MUST be passed to the actual consumer and MUST NOT be held by the caller.
+        // This is due to the fact that deiniting the sequence is used as part of a trigger to
+        // terminate the underlying source.
+        let acknowledgementsSourceAndSequence = NIOAsyncSequenceProducer.makeSequence(
+            of: Acknowledgement.self,
+            backPressureStrategy: NoBackPressure(),
+            delegate: NoDelegate()
+        )
+        self.acknowledgementsSource = acknowledgementsSourceAndSequence.source
+        self.acknowledgements = AcknowledgedMessagesAsyncSequence(
+            wrappedSequence: acknowledgementsSourceAndSequence.sequence
+        )
+
+        // Create a new reference to self that is then passed
+        // to the KafkaConfig as an opaque pointer.
+        // To avoid memory leaks, this reference is evantually
+        // removed in the shutdownGracefully method.
+        self.referenceToSelf = Unmanaged.passRetained(self).toOpaque()
+
+        self.config.setDeliveryReportCallback(callback: self.deliveryReportCallback)
+        self.config.setOpaque(opaque: self.referenceToSelf)
+
+        self.client = try KafkaClient(type: .producer, config: self.config, logger: self.logger)
+        self.kafkaHandle = self.client.kafkaHandle
+
+        // Poll Kafka every millisecond
+        self.pollTask = Task { [kafkaHandle] in
+            while !Task.isCancelled {
+                rd_kafka_poll(kafkaHandle, 0)
+                try? await Task.sleep(nanoseconds: 1_000_000)
+            }
+        }
+    }
+
+    /// Method to shutdown the ``KafkaProducer``.
+    ///
+    /// This method flushes any buffered messages and waits until a callback is received for all of them.
+    /// Afterwards, it shuts down the connection to Kafka and cleans any remaining state up.
+    public func shutdownGracefully() throws {
+        switch self.state {
+        case .started:
+            self.state = .shuttingDown
+            self._shutDownGracefully()
+        case .shuttingDown, .shutDown:
+            return
+        }
+    }
+
+    private func _shutDownGracefully() {
+        // Wait 10 seconds for outstanding messages to be sent and callbacks to be called
+        rd_kafka_flush(self.kafkaHandle, 10000)
```

Comment:
Do we know that no further work can come in after this point?

Reply:
The 10 seconds were rather chosen arbitrarily — we could potentially have something like an outgoing message queue and wait until all outstanding messages have been acknowledged. However, this could take forever.